src/macaron/artifact/maven.py

```diff
@@ -0,0 +1,145 @@
+# Copyright (c) 2024 - 2024, Oracle and/or its affiliates. All rights reserved.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.
+
+"""This module declares types and utilities for Maven artifacts."""
+
+import re
+from dataclasses import dataclass
+from enum import Enum
+from typing import NamedTuple, Self
+
+from packageurl import PackageURL
+
+
+class _MavenArtifactType(NamedTuple):
+    filename_pattern: str
+    purl_qualifiers: dict[str, str]
+
+
+class MavenArtifactType(_MavenArtifactType, Enum):
+    """Maven artifact types that Macaron supports.
+
+    For reference, see:
+    - https://maven.apache.org/ref/3.9.6/maven-core/artifact-handlers.html
+    - https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#maven
+    """
+
+    # Enum with custom value type.
+    # See https://docs.python.org/3.10/library/enum.html#others.
+    JAR = _MavenArtifactType(
+        filename_pattern="{artifact_id}-{version}.jar",
+        purl_qualifiers={"type": "jar"},
+    )
+    POM = _MavenArtifactType(
+        filename_pattern="{artifact_id}-{version}.pom",
+        purl_qualifiers={"type": "pom"},
+    )
+    JAVADOC = _MavenArtifactType(
+        filename_pattern="{artifact_id}-{version}-javadoc.jar",
+        purl_qualifiers={"type": "javadoc"},
+    )
+    JAVA_SOURCE = _MavenArtifactType(
+        filename_pattern="{artifact_id}-{version}-sources.jar",
+        purl_qualifiers={"type": "sources"},
+    )
```

Comment:
I wonder if the `type` key within `purl_qualifiers` maps directly to the `type` column in the Default Artifact Handlers Reference table [here](https://maven.apache.org/ref/3.9.6/maven-core/artifact-handlers.html)?

Reply:
Yes, that's correct. As stated in the [Maven section of the PackageURL doc](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#maven) which I have left a link to:
> Known qualifiers keys are: classifier and type as defined in the POM documentation.

Hope this clarifies.

Reply:
Hmmm right, I have some questions regarding this:
- Why is the `type` field for `JAVA_SOURCE` is `sources` while in the `type` column [here](https://maven.apache.org/ref/3.9.6/maven-core/artifact-handlers.html), it's `java-source`.
> Known qualifiers keys are: classifier and type as defined in the POM documentation.
- If it's mentioned that there are two type of qualifiers `type` and `classifier` for a Maven PURL. I wonder why for `JAVADOC` and `JAVA_SOURCE` we only have one quantifier. Shouldn't it be something like this :thinking: ?
```
    JAVADOC = _MavenArtifactType(
        filename_pattern="{artifact_id}-{version}-javadoc.jar",
        purl_qualifiers={"type": "javadoc", "classifier": "javadoc"},
    )
    JAVA_SOURCE = _MavenArtifactType(
        filename_pattern="{artifact_id}-{version}-sources.jar",
        purl_qualifiers={"type": "java-source", "classifier": "sources},
    )
```
In additions, we perform a dictionary comparison here 
https://github.com/oracle/macaron/blob/7fa98d6e8118dca219f927cfaf123a368c54fd54/src/macaron/artifact/maven.py#L100-L102
It would mean that we expect the qualifier from the input package url must only contains one key-value which is `{ "type": "<value>" }`.
From the examples given [here](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#maven). It seems to me that a Maven package URL can have both `type` and/or `classifier` quantifiers. 
Are we only supporting the package URL with only `type` quantifier in this PR? :thinking:

Reply:
Thanks for pointing out the issue with the `type` qualifier of `java-source`. You are correct that it should be `java-source` and not `sources`. Fixed in [ebf8667](https://github.com/oracle/macaron/pull/685/commits/ebf866729da04f25871a8125519eb665f2dcee32).

I have decided to only support the `type` qualifier for now. I added a comment to clarify that in the same commit. The `type` qualifier is enough for uniquely identifying the artifact type of each artifact while the `classifier` qualifier does not have this property. In addition, supporting `classifier` will make the implementation quite a bit more complex, while there is no concrete usage for it for now, so I figure supporting only the `type` qualifier at the moment is probably sufficient.

Reply:
Sounds good to me. Thanks.